rmw_cyclonedds_cpp/src/serdata.cpp

```diff
@@ -685,11 +749,351 @@ static std::string get_type_name(const char * type_support_identifier, void * ty
   }
 }
 
+
+extern "C" dds_dynamic_type_descriptor_t get_dynamic_type_descriptor_prim(
+  dds_dynamic_type_kind_t kind, const char *name, uint32_t num_bounds, const uint32_t *bounds, dds_dynamic_type_kind_t type)
+{
+  return{kind,name,{},{},num_bounds,bounds,DDS_DYNAMIC_TYPE_SPEC_PRIM(type),{},};
+}
+
+extern "C" dds_dynamic_type_descriptor_t get_dynamic_type_descriptor(
+  dds_dynamic_type_kind_t kind, const char *name, uint32_t num_bounds, const uint32_t *bounds, dds_dynamic_type_t type)
+{
+  return{kind,name,{},{},num_bounds,bounds,DDS_DYNAMIC_TYPE_SPEC(type),{},};
+}
+
+extern "C" dds_dynamic_member_descriptor_t get_dynamic_member_descriptor_prim(
+  dds_dynamic_type_kind_t type, const char *name)
+{
+  return DDS_DYNAMIC_MEMBER_PRIM(type, name);
+}
+
+extern "C" dds_dynamic_member_descriptor_t get_dynamic_member_descriptor(
+  dds_dynamic_type_t ddt, const char *name)
+{
+  return DDS_DYNAMIC_MEMBER(ddt, name);
+}
+
+
+template<typename MemberType>
+static void dynamic_type_add_array(
+  dds_dynamic_type_t * dstruct, dds_entity_t dds_ppant, const MemberType * member, const dds_dynamic_type_kind_t type)
+{
+  dds_dynamic_type_t ddt;
+
+  if (member->array_size_) 
+  { 
+    if (!member->is_upper_bound_)
+    {
+      ddt = dds_dynamic_type_create(dds_ppant, 
+        get_dynamic_type_descriptor_prim(DDS_DYNAMIC_ARRAY, member->name_, 1, (uint32_t*)(&member->array_size_), type));
```

Comment:
I believe the actual type of `array_size_` is `size_t`, so casting it to a pointer to `uint32_t` will break things quite badly on big-endian 64-bit machines.

I've checked and `dds_dynamic_type_create` in turn calls `ddsi_dynamic_type_create_array`, which copies the data. Given that it is a single-element array, changing it to:
```
uint32_t array_size = static_cast<uint32_t>(member->array_size_);
ddt = dds_dynamic_type_create(dds_ppant, 
        get_dynamic_type_descriptor_prim(DDS_DYNAMIC_ARRAY, member->name_, 1, &array_size_, type));
```
should solve that.

Reply:
Apply your suggestions.